In [15]:
from pydataset import data
import pandas as pd
import numpy as np

## Part I Below

In [135]:
import env

In [136]:
import os

In [ ]:
#db_url_function
def get_db_url(db, env_file=os.path.exists('env.py')):
    '''
    return a formatted string containing username, password,
    host and database for connecting to the mySQL server and 
    the database indicated.
    env_file checks to see if the env.py exists in cwd
    
    '''
    if env_file:
        username, password, host = (env.username, env.password, env.host)
        return f'mysql+pymysql://{username}:{password}@{host}/{db}'
    else:
        return 'hello, you need a username and password to get in!'

In [83]:
url=get_db_url('employees')

In [84]:
query = 'SELECT * FROM employees LIMIT 10'

In [85]:
#employees database
employees_df = pd.read_sql(query, url)
employees_df.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

#### Database Typos

In [86]:
#db_url_function typos
def get_db_url(db, env_file=os.path.exists('env.py')):
    '''
    return a formatted string containing username, password,
    host and database for connecting to the mySQL server and 
    the database indicated.
    env_file checks to see if the env.py exists in cwd
    
    '''
    if env_file:
        usernames, password, host = (env.username, env.password, env.host)
        return f'mysql+pymysql://{username}:{password}@{host}/{db}'
    else:
        return 'hello, you need a username and password to get in!'



NameError: name 'username' is not defined

// The error is "usernames" instead of "username"

In [88]:
#db_url_function typos
def get_db_url(db, env_file=os.path.exists('env.py')):
    '''
    return a formatted string containing username, password,
    host and database for connecting to the mySQL server and 
    the database indicated.
    env_file checks to see if the env.py exists in cwd
    
    '''
    if env_file:
        usernames, password, host = (env.username, env.password, env.host)
        return f'mysql+pymysql://{noether_1234}:{password}@{host}/{db}'
    else:
        return 'hello, you need a username and password to get in!'



//Replaced {username} with {noether_1234}

In [90]:
url=get_db_url('employees')

NameError: name 'noether_2029' is not defined

//Error messaged occured due to incorrect formating

### Two seperate DataFrames

In [101]:
query = 'SELECT * FROM employees'

In [103]:
#employees database
employees_df = pd.read_sql(query, url)
employees_df.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [107]:
query = '''
SELECT * 
FROM titles
'''

In [108]:
#titles database
titles_df = pd.read_sql(query, url)
titles_df.head()

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

#### How many rows and columns do you have in each DataFrame? Is that what you expected?

In [105]:
employees_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [109]:
titles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


//There are 30024 rows and 6 columns in employees. 

//There are 443308 rows and 4 columns in titles.

// Yes

### Summary statistics

In [110]:
employees_df.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [111]:
titles_df.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

#### Unique Titles in Titles DataFrame.

In [116]:
titles_df.nunique().nsmallest(1)

title    7
dtype: int64

In [127]:
query = '''
SELECT to_date 
FROM titles
ORDER BY to_date ASC
limit 1;
'''

In [128]:
titles_df = pd.read_sql(query, url)

In [129]:
titles_df

to_date
0  1985-03-01

// 1985-03-01 is the oldest date. 

In [130]:
query = '''
SELECT to_date 
FROM titles
WHERE to_date <= CURDATE()
ORDER BY to_date DESC
limit 1;
'''

In [132]:
titles_df = pd.read_sql(query, url)

In [133]:
titles_df

to_date
0  2002-08-01

// The most recent date is 2002-08-01.

## Conclusion of Part I

# Part II Below

In [5]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [6]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

In [8]:
(users.merge(roles,                  #Right Join
            left_on='role_id', 
            right_on='id', 
            how='right', 
            indicator=True
            )
)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

In [9]:
(users.merge(roles,                  #Outer Join
            left_on='role_id', 
            right_on='id', 
            how='outer', 
            indicator=True
            )
)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

In [233]:
(users.merge(roles,                  #dropped foreign keys
            left_on='role_id', 
            right_on='id', 
            how='outer', 
            indicator=True
            )
 .drop(columns=['role_id', 'id_y'])
)

id_x name_x     name_y      _merge
0   1.0    bob      admin        both
1   2.0    joe     author        both
2   3.0  sally   reviewer        both
3   4.0   adam   reviewer        both
4   5.0   jane        NaN   left_only
5   6.0   mike        NaN   left_only
6   NaN    NaN  commenter  right_only

#### MPG Dataset

In [235]:
mpg = data('mpg')


In [236]:
mpg.shape

(234, 11)

In [237]:
mpg.groupby('manufacturer').describe().T

manufacturer         audi    chevrolet        dodge         ford        honda  \
displ count     18.000000    19.000000    37.000000    25.000000     9.000000   
      mean       2.544444     5.063158     4.378378     4.536000     1.711111   
      std        0.673203     1.370406     0.867991     0.540740     0.145297   
      min        1.800000     2.400000     2.400000     3.800000     1.600000   
      25%        2.000000     4.450000     3.800000     4.000000     1.600000   
      50%        2.800000     5.300000     4.700000     4.600000     1.600000   
      75%        3.025000     5.850000     4.700000     4.600000     1.800000   
      max        4.200000     7.000000     5.900000     5.400000     2.000000   
year  count     18.000000    19.000000    37.000000    25.000000     9.000000   
      mean    2003.500000  2004.684211  2004.108108  2002.600000  2003.000000   
      std        4.630462     4.460352     4.520225     4.500000     4.743416   
      min     1999.000000  1999.000000  1999.000000  1999.000000  1999.000000   
      25%     1999.000000  1999.000000  1999.000000  1999.000000  1999.000000   
      50%     2003.500000  2008.000000  2008.000000  1999.000000  1999.000000   
      75%     2008.000000  2008.000000  2008.000000  2008.000000  2008.000000   
      max     2008.000000  2008.000000  2008.000000  2008.000000  2008.000000   
cyl   count     18.000000    19.000000    37.000000    25.000000     9.000000   
      mean       5.222222     7.263158     7.081081     7.200000     4.000000   
      std        1.215370     1.367971     1.115008     1.000000     0.000000   
      min        4.000000     4.000000     4.000000     6.000000     4.000000   
      25%        4.000000     7.000000     6.000000     6.000000     4.000000   
      50%        6.000000     8.000000     8.000000     8.000000     4.000000   
      75%        6.000000     8.000000     8.000000     8.000000     4.000000   
      max        8.000000     8.000000     8.000000     8.000000     4.000000   
cty   count     18.000000    19.000000    37.000000    25.000000     9.000000   
      mean      17.611111    15.000000    13.135135    14.000000    24.444444   
      std        1.974511     2.924988     2.485091     1.914854     1.943651   
      min       15.000000    11.000000     9.000000    11.000000    21.000000   
      25%       16.000000    13.500000    11.000000    13.000000    24.000000   
      50%       17.500000    15.000000    13.000000    14.000000    24.000000   
      75%       18.750000    16.500000    15.000000    15.000000    25.000000   
      max       21.000000    22.000000    18.000000    18.000000    28.000000   
hwy   count     18.000000    19.000000    37.000000    25.000000     9.000000   
      mean      26.444444    21.894737    17.945946    19.360000    32.555556   
      std        2.175322     5.108759     3.574182     3.327662     2.554952   
      min       23.000000    14.000000    12.000000    15.000000    29.000000   
      25%       25.000000    17.000000    16.000000    17.000000    32.000000   
      50%       26.000000    23.000000    17.000000    18.000000    32.000000   
      75%       27.750000    26.000000    21.000000    22.000000    34.000000   
      max       31.000000    30.000000    24.000000    26.000000    36.000000   

manufacturer      hyundai         jeep   land rover       lincoln  \
displ count     14.000000     8.000000     4.000000  3.000000e+00   
      mean       2.428571     4.575000     4.300000  5.400000e+00   
      std        0.364646     1.015241     0.258199  1.087792e-15   
      min        2.000000     3.000000     4.000000  5.400000e+00   
      25%        2.100000     3.925000     4.150000  5.400000e+00   
      50%        2.400000     4.700000     4.300000  5.400000e+00   
      75%        2.650000     4.950000     4.450000  5.400000e+00   
      max        3.300000     6.100000     4.600000  5.400000e+00   
year  count     14.000000     8.000000     4.000000  3.00

In [238]:
mpg.nunique(axis=0).head(1) #15 manufacturers

manufacturer    15
dtype: int64

In [239]:
mpg.nunique().head(2)    #38 models  

manufacturer    15
model           38
dtype: int64

In [240]:
mpg.assign(mileage_difference = mpg.hwy-mpg.cty)       #mileage_difference

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

In [246]:
mpg.assign(average_mileage = (mpg.hwy+mpg.cty)/2)    #Average_mileage

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class is_automatic  average_mileage  
1    compact    Automatic             23.5  
2    compact       Manual             25.0  
3    compact       Manual             25.5  
4    compact    Automatic             25.5  
5    compact    Automatic             21.0  
..       ...          ...              ...  
230  midsize    Automatic             23.5  
231  midsize       Manual             25.0  
232  midsize    Automatic             21.0  
233  midsize       Manual             22.0  
234  midsize    Automatic             21.5  

[234 rows x 13 columns]

In [242]:
mpg['is_automatic']=mpg.trans.str.contains('auto')==True

In [243]:
mpg['is_automatic'] = mpg['is_automatic'].replace({True: 'Automatic', False: 'Manual'})

In [244]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class is_automatic  
1    compact    Automatic  
2    compact       Manual  
3    compact       Manual  
4    compact    Automatic  
5    compact    Automatic  
..       ...          ...  
230  midsize    Automatic  
231  midsize       Manual  
232  midsize    Automatic  
233  midsize       Manual  
234  midsize    Automatic  

[234 rows x 12 columns]

In [250]:
mpg.assign(average_mileage = (mpg.hwy+mpg.cty)/2).groupby('manufacturer')['average_mileage'].agg('mean').sort_values(ascending=False)

manufacturer
honda         28.500000
volkswagen    25.074074
hyundai       22.750000
subaru        22.428571
audi          22.027778
toyota        21.720588
pontiac       21.700000
nissan        21.346154
chevrolet     18.447368
ford          16.680000
mercury       15.625000
jeep          15.562500
dodge         15.540541
lincoln       14.166667
land rover    14.000000
Name: average_mileage, dtype: float64

In [253]:
mpg['is_automatic'] = mpg['is_automatic'].replace({True: 'Automatic', False: 'Manual'}).groupby('is_automatic')['average_mileage'].agg('mean')

KeyError: 'is_automatic'

## Conclusion of Part II

### Part III Below

In [268]:
def get_db_url(db, env_file=os.path.exists('env.py')):
    '''
    return a formatted string containing username, password,
    host and database for connecting to the mySQL server and 
    the database indicated.
    env_file checks to see if the env.py exists in cwd
    
    '''
    if env_file:
        username, password, host = (env.username, env.password, env.host)
        return f'mysql+pymysql://{username}:{password}@{host}/{db}'
    else:
        return 'hello, you need a username and password to get in!'

In [269]:
url=get_db_url('chipotle')

In [270]:
query = '''
SELECT item_price
FROM orders;
'''

### Chipotle database

In [259]:
#employees database
chipotle_df = pd.read_sql(query, url)

In [157]:
chipotle_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   id                  4622 non-null   int64 
 1   order_id            4622 non-null   int64 
 2   quantity            4622 non-null   int64 
 3   item_name           4622 non-null   object
 4   choice_description  4622 non-null   object
 5   item_price          4622 non-null   object
dtypes: int64(3), object(3)
memory usage: 216.8+ KB


In [217]:
chipotle_df.head()

item_price
0     $2.39 
1     $3.39 
2     $3.39 
3     $2.39 
4    $16.98

In [231]:
chipotle_df.item_price.str.replace('$', '').astype(float)

/var/folders/51/1y9bskdx161fnwxyf68jcl540000gn/T/ipykernel_4562/377780417.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  chipotle_df.item_price.str.replace('$', '').astype(float).sum()


34500.16

In [271]:
chipotle_df.groupby('order_id')['item_price'].agg('sum')

KeyError: 'order_id'

In [263]:
chipotle_df['item_name'].count()

KeyError: 'item_name'

In [265]:
chipotle_df.groupby('item_name')['id'].count().nlargest(3)

KeyError: 'item_name'

In [266]:
chipotle_df('item_name')['item_price'].agg('sum').nlargest(1)

TypeError: 'DataFrame' object is not callable

In [267]:
employees.head()

NameError: name 'employees' is not defined

In [273]:
employees.merge(titles, on='emp_no')

NameError: name 'employees' is not defined

In [272]:
emp_title = employees.merge(titles, on='emp_no')

NameError: name 'employees' is not defined

In [274]:
emp_title.groupby('title')['from_date'].agg('max')

NameError: name 'emp_title' is not defined

In [ ]:
cross_query = '''
SELECT t.title, d.dept_name
FROM dept_emp
JOIN titles t USING(emp_no)
JOIN departments d USING(dept_no)
WHERE t.to_date > CURDATE();
'''

In [ ]:
cross_url = get_db_url(host, username, password, 'employees')

In [ ]:
cross_url = get_db_url(cross_query, cross_url)

In [ ]:
df_cross = pd.read_sql(cross_query, cross_url)

df_cross.head()

In [ ]:
df_cross.shape

In [ ]:
pd.crossbar(df_cross['title'], df_cross['dept_name']).T